rfd/0083-machine-id-host-certs.md

```diff
@@ -0,0 +1,198 @@
+
+   We should provide (and document) a sane rule that ensures regular SSH host
+   cert rules are followed, alongside users' own requirements (e.g. DNS suffix).
+
+[`GenerateHostCert()`]: https://github.com/gravitational/teleport/blob/82c520c8183553f310459c3b4a96b70065ee268a/lib/auth/auth_with_roles.go#L2139
+[`SSHSession`]: https://github.com/gravitational/teleport/blob/ab12ad33d9b3143baa5dc1a0c236cb6ed7645f10/lib/services/parser.go#L183
+
+### Issuing short-lived certificates
+
+SSH host certificates issued today via `tctl auth sign` [do not have an
+expiration date][date]. While we could preserve this behavior, any process that
+continually produces certificates with an infinite TTL seems problematic.
+
+Our preference is to generate similarly short-lived certificates as we do for
+other bot credentials; this has certain (minor) caveats as explored in the UX
+section below.
+
+[date]: https://github.com/gravitational/teleport/blob/ab12ad33d9b3143baa5dc1a0c236cb6ed7645f10/tool/tctl/common/auth_command.go#L426
+
+### UX
+
+We have several options for issuing host certificates. Factors to consider
+include:
+1. Certificate lifespan. Do these certificates need to expire every hour? Our
+   documented advice today generates certificates with no expiration date
+   specified.
+2. Does issuing them every hour create additional UX problems of its own? We
+   found this is largely safe with certain caveats (see below).
+3. Does `sshd` gracefully reload these certificates? We found that sshd re-reads
+   the certificate file for each incoming connection, so no reload is necessary
+   unless the `sshd_config` file itself has changed (e.g. changed the path to
+   the host cert / key).
+
+#### Preferred implementation: Config Template
+
+Config Templates in Machine ID are refreshed each iteration of the bots usual
+renewal loop (20min by default). This would be the most straightforward
+integration approach and we'd approach it like we do other special cert types
+for apps / databases / k8s clusters.
+
+Users would request a host cert via the config file:
+
+```yaml
+destinations:
+  - directory: /opt/machine-id
+    configs:
+      - host_cert:
+          principals: [foo.example.com]
+```
+
+While rendering the config template, the bot would call `GenerateHostCert()`,
+format the resulting certificate and key appropriately, and write it to disk.
+A new certificate would be written at startup, then approximately every 20
+minutes while running normally.
+
+Per testing, OpenSSH re-reads certificate files on demand, so no additional work
+is needed to make use of these certificates once paths are configured.
+Additionally, sshd does play nice with `tbot init`'s ACL implementation, and
+does not care about file permissions unless they're owned by `root`.
+
+A downside to this approach is that permissions errors (for example, if the
+`where` predicate doesn't match) will be reported as errors but will not crash
```

Comment:
Should we emit audit events for these errors?

Reply:
It is logged today (in debug mode only) but not as an audit event. I think this is standard Teleport RBAC behavior? I don't see any events emitted for any failed RBAC checks, surprisingly.

Reply:
It feels like we certainly should be auditing failed RBAC checks, especially for generating certificates ! If you feel its out of scope for this, can we at least raise an issue to add an audit event for this ?

Reply:
We do have rule actions which can produce logs (though not audit logs) on the off chance someone wants that for a particular rule. That might be an okay way forward, maybe.

Reply:
Yes, perhaps not as useful for customers who are only ingesting the audit log into their SIEM, but probably the best we can do for now as part of this work.